[CBRD-25945] fix coredump

mhoh3963 opened the pull request:
http://jira.cubrid.org/browse/CBRD-25945

- change assert condition of session_state_destroy()
- modify to check if tdes is NULL in the  logtb_set_current_user_active()

mhoh3963 committed:
[CBRD-25945] fix coredump

http://jira.cubrid.org/browse/CBRD-25945

- change assert condition of session_state_destroy()
- modify to check if tdes is NULL in the  logtb_set_current_user_active()

beyondykk9 commented on src/transaction/log_tran_table.c:
```diff
@@ -2083,7 +2083,10 @@ logtb_set_current_user_active (THREAD_ENTRY * thread_p, bool is_user_active)
   tdes = LOG_FIND_TDES (tran_index);
 
-  tdes->is_user_active = is_user_active;
+  if (!tdes)
```
```suggestion
  if (tdes)
```

beyondykk9 replied:
tdes가 NULL이 아니어야 하지 않나요?

mhoh3963 committed:
Update src/transaction/log_tran_table.c

Co-authored-by: Byungwook Kim <bwkim@cubrid.com>

hornetmj commented on src/session/session.c:
```diff
@@ -773,7 +773,7 @@ session_state_destroy (THREAD_ENTRY * thread_p, const SESSION_ID id, bool is_kee
 
 #if defined (SERVER_MODE)
-  assert (session_p->ref_count > 0);
+  assert (session_p->ref_count >= 0);
```
1. session_state_destroy() 함수는 정상적인 세션 종료 요청 시 호출되는 함수로 session_p->ref_count == 0을 허용하는 조건은 적합해 보이지 않습니다. 우회를 위한 것이라면 관련 정보를 주석으로 남겨두시면 좋을 것 같습니다.
2. session_state_decrease_ref_count()가 호출되는 경우 해당 함수 내에서 state_p->ref_count가 -1이 되어 다시 assert에 걸릴 수 있습니다.

mhoh3963 replied:
정확히 ref_count=0이면서 session이 등록된 경우를 몰라서 assert 대신 방어 코드를 넣는 것으로 수정하겠습니다.

mhoh3963 committed:
- modify db_end_session() as the following
  * if session_id is DB_EMPTY_SESSION, then it returns immediately.
  * if keep_session is false and csession_end_session() is success, session_id is set to DB_EMPTY_SESSION.
- There is a place where the ref_count is reduced without locking session_p in another function of the server, so there may be a case where the ref_count is 0 in the session_state_destroy() function. So the assert is removed and a defensive code is added.

mhoh3963 committed:
Merge branch 'cbrd-25945' of github.com:mhoh3963/cubrid into cbrd-25945

mhoh3963 commented:
@beyondykk9  @hornetmj 
3가지 방어 코드를 추가했습니다.
- db_end_session() 에서 empty_session인 경우 바로 반환하게 수정
- db_end_session()에서 csession_end_session() 성공적으로 처리되고 keep_session이 false일 경우 session_id를 empty_session으로 변경
- server code중 session_p를 lock 없이 ref_count만 줄이는 함수들이 있어, ref_count가 0인 경우가 발생할 가능성이 있어 assert를 제거하고 방어코드를 추가

beyondykk9 commented on src/compat/db_admin.c:
```diff
@@ -1052,14 +1052,19 @@ db_enable_modification (void)
  * NOTE: This function ends the session identified by 'db_Session_id'
  */
-static int is_doing_end_session = -1;
+static SESSION_ID is_doing_end_session = -1;
```
SESSION_ID는 unsigned int로 음수 값을 가질 수 없습니다.

mhoh3963 replied:
리뷰, 감사합니다.
수정해서 반영하겠습니다.

mhoh3963 committed:
- fix miss data type
